<a href="https://colab.research.google.com/github/bema1015/ProjDataEng/blob/main/Projet2/projet2_dev.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Contexte

Mark, votre manager, vous a convié à une réunion pour vous présenter le
projet d’expansion à l’international de l’entreprise. Il vous confie une
première mission d’analyse exploratoire, pour déterminer si les données sur
l’éducation de la banque mondiale permettent d’informer le projet
d’expansion.
Voici les différentes questions que Mark aimerait explorer, que vous avez
notées durant la réunion :
*   Quels sont les pays avec un fort potentiel de clients pour nos services ?
*   Pour chacun de ces pays, quelle sera l’évolution de ce potentiel de clients ?
*   Dans quels pays l'entreprise doit-elle opérer en priorité ?

# Jeu de donnée

Issu de la Banque Mondiale :
https://datacatalog.worldbank.org/dataset/education-statistics
Disponible en telchargment direct à https://s3-eu-west-1.amazonaws.com/static.ocstatic.com/prod/courses/files/Parcours_data_scientist/Projet+-
+Donn%C3%A9es+%C3%A9ducatives/Projet+Python_Dataset_Edstats_csv.zip
Le fichier zip contient 5 fichiers csv :
* EdStatsData.csv
* EdStatsFootNote.csv
* EdStatsSeries.csv
* EdStatCountry.csv
* EdStatsCountry-Series.csv


In [3]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns
from google.colab import drive
drive.mount('/content/drive')

Mounted at /content/drive


#Chargement & Aperçu

Objectif :
* Charger les données

Résumé :
* EdStatsData.csv = EDSDATA
* EdStatsFootNote.csv = EDSFootNote
* EdStatsSeries.csv = EDSSeries
* EdStatCountry.csv = EDSCountry
* EdStatsCountry-Series.csv = EDSCountrySeries

In [4]:
EDSData = pd.read_csv('/content/drive/MyDrive/Colab Notebooks/Projet+Python_Dataset_Edstats_csv/EdStatsData.csv')
EDSData.head()

Country Name Country Code  \
0   Arab World          ARB   
1   Arab World          ARB   
2   Arab World          ARB   
3   Arab World          ARB   
4   Arab World          ARB   

                                      Indicator Name  Indicator Code  \
0  Adjusted net enrolment rate, lower secondary, ...      UIS.NERA.2   
1  Adjusted net enrolment rate, lower secondary, ...    UIS.NERA.2.F   
2  Adjusted net enrolment rate, lower secondary, ...  UIS.NERA.2.GPI   
3  Adjusted net enrolment rate, lower secondary, ...    UIS.NERA.2.M   
4  Adjusted net enrolment rate, primary, both sex...     SE.PRM.TENR   

        1970       1971       1972       1973       1974      1975  ...  2060  \
0        NaN        NaN        NaN        NaN        NaN       NaN  ...   NaN   
1        NaN        NaN        NaN        NaN        NaN       NaN  ...   NaN   
2        NaN        NaN        NaN        NaN        NaN       NaN  ...   NaN   
3        NaN        NaN        NaN        NaN        NaN       NaN  ...   NaN   
4  54.822121  54.894138  56.209438  57.267109  57.991138  59.36554  ...   NaN   

   2065  2070  2075  2080  2085  2090  2095  2100  Unnamed: 69  
0   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN          NaN  
1   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN          NaN  
2   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN          NaN  
3   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN          NaN  
4   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN          NaN  

[5 rows x 70 columns]

In [5]:
EDSCountrySeries = pd.read_csv('/content/drive/MyDrive/Colab Notebooks/Projet+Python_Dataset_Edstats_csv/EdStatsCountry-Series.csv')
EDSCountrySeries.head()

CountryCode         SeriesCode  \
0         ABW        SP.POP.TOTL   
1         ABW        SP.POP.GROW   
2         AFG        SP.POP.GROW   
3         AFG  NY.GDP.PCAP.PP.CD   
4         AFG        SP.POP.TOTL   

                                         DESCRIPTION  Unnamed: 3  
0  Data sources : United Nations World Population...         NaN  
1  Data sources: United Nations World Population ...         NaN  
2  Data sources: United Nations World Population ...         NaN  
3                 Estimates are based on regression.         NaN  
4  Data sources : United Nations World Population...         NaN

In [6]:
EDSCountry = pd.read_csv('/content/drive/MyDrive/Colab Notebooks/Projet+Python_Dataset_Edstats_csv/EdStatsCountry.csv')
EDSCountry.head()

Country Code   Short Name   Table Name                     Long Name  \
0          ABW        Aruba        Aruba                         Aruba   
1          AFG  Afghanistan  Afghanistan  Islamic State of Afghanistan   
2          AGO       Angola       Angola   People's Republic of Angola   
3          ALB      Albania      Albania           Republic of Albania   
4          AND      Andorra      Andorra       Principality of Andorra   

  2-alpha code   Currency Unit  \
0           AW   Aruban florin   
1           AF  Afghan afghani   
2           AO  Angolan kwanza   
3           AL    Albanian lek   
4           AD            Euro   

                                       Special Notes  \
0  SNA data for 2000-2011 are updated from offici...   
1  Fiscal year end: March 20; reporting period fo...   
2  April 2013 database update: Based on IMF data,...   
3                                                NaN   
4                                                NaN   

                      Region          Income Group WB-2 code  ...  \
0  Latin America & Caribbean  High income: nonOECD        AW  ...   
1                 South Asia            Low income        AF  ...   
2         Sub-Saharan Africa   Upper middle income        AO  ...   
3      Europe & Central Asia   Upper middle income        AL  ...   
4      Europe & Central Asia  High income: nonOECD        AD  ...   

            IMF data dissemination standard  \
0                                       NaN   
1  General Data Dissemination System (GDDS)   
2  General Data Dissemination System (GDDS)   
3  General Data Dissemination System (GDDS)   
4                                       NaN   

                            Latest population census  \
0                                               2010   
1                                               1979   
2                                               1970   
3                                               2011   
4  2011. Population figures compiled from adminis...   

                             Latest household survey  \
0                                                NaN   
1  Multiple Indicator Cluster Survey (MICS), 2010/11   
2               Malaria Indicator Survey (MIS), 2011   
3       Demographic and Health Survey (DHS), 2008/09   
4                                                NaN   

   Source of most recent Income and expenditure data  \
0                                                NaN   
1            Integrated household survey (IHS), 2008   
2            Integrated household survey (IHS), 2008   
3  Living Standards Measurement Study Survey (LSM...   
4                                                NaN   

  Vital registration complete Latest agricultural census  \
0                         Yes                        NaN   
1                         NaN                    2013/14   
2                         NaN                       2015   
3                         Yes                       2012   
4                         Yes                        NaN   

  Latest industrial data Latest trade data Latest water withdrawal data  \
0                    NaN            2012.0                          NaN   
1                    NaN            2012.0                         2000   
2                    NaN               NaN                         2005   
3                 2010.0            2012.0                         2006   
4                    NaN            2006.0                          NaN   

  Unnamed: 31  
0         NaN  
1         NaN  
2         NaN  
3         NaN  
4         NaN  

[5 rows x 32 columns]

In [7]:
EDSFootNote = pd.read_csv('/content/drive/MyDrive/Colab Notebooks/Projet+Python_Dataset_Edstats_csv/EdStatsFootNote.csv')
EDSFootNote.head()

CountryCode      SeriesCode    Year          DESCRIPTION  Unnamed: 4
0         ABW  SE.PRE.ENRL.FE  YR2001  Country estimation.         NaN
1         ABW  SE.TER.TCHR.FE  YR2005  Country estimation.         NaN
2         ABW  SE.PRE.TCHR.FE  YR2000  Country estimation.         NaN
3         ABW  SE.SEC.ENRL.GC  YR2004  Country estimation.         NaN
4         ABW     SE.PRE.TCHR  YR2006  Country estimation.         NaN

In [8]:
EDSSeries = pd.read_csv('/content/drive/MyDrive/Colab Notebooks/Projet+Python_Dataset_Edstats_csv/EdStatsSeries.csv')
EDSSeries.head()
#

Series Code       Topic  \
0  BAR.NOED.1519.FE.ZS  Attainment   
1     BAR.NOED.1519.ZS  Attainment   
2  BAR.NOED.15UP.FE.ZS  Attainment   
3     BAR.NOED.15UP.ZS  Attainment   
4  BAR.NOED.2024.FE.ZS  Attainment   

                                      Indicator Name  \
0  Barro-Lee: Percentage of female population age...   
1  Barro-Lee: Percentage of population age 15-19 ...   
2  Barro-Lee: Percentage of female population age...   
3  Barro-Lee: Percentage of population age 15+ wi...   
4  Barro-Lee: Percentage of female population age...   

                                    Short definition  \
0  Percentage of female population age 15-19 with...   
1  Percentage of population age 15-19 with no edu...   
2  Percentage of female population age 15+ with n...   
3  Percentage of population age 15+ with no educa...   
4  Percentage of female population age 20-24 with...   

                                     Long definition  Unit of measure  \
0  Percentage of female population age 15-19 with...              NaN   
1  Percentage of population age 15-19 with no edu...              NaN   
2  Percentage of female population age 15+ with n...              NaN   
3  Percentage of population age 15+ with no educa...              NaN   
4  Percentage of female population age 20-24 with...              NaN   

  Periodicity Base Period Other notes Aggregation method  ...  \
0         NaN         NaN         NaN                NaN  ...   
1         NaN         NaN         NaN                NaN  ...   
2         NaN         NaN         NaN                NaN  ...   
3         NaN         NaN         NaN                NaN  ...   
4         NaN         NaN         NaN                NaN  ...   

  Notes from original source  General comments  \
0                        NaN               NaN   
1                        NaN               NaN   
2                        NaN               NaN   
3                        NaN               NaN   
4                        NaN               NaN   

                                              Source  \
0  Robert J. Barro and Jong-Wha Lee: http://www.b...   
1  Robert J. Barro and Jong-Wha Lee: http://www.b...   
2  Robert J. Barro and Jong-Wha Lee: http://www.b...   
3  Robert J. Barro and Jong-Wha Lee: http://www.b...   
4  Robert J. Barro and Jong-Wha Lee: http://www.b...   

  Statistical concept and methodology Development relevance  \
0                                 NaN                   NaN   
1                                 NaN                   NaN   
2                                 NaN                   NaN   
3                                 NaN                   NaN   
4                                 NaN                   NaN   

  Related source links Other web links  Related indicators  License Type  \
0                  NaN             NaN                 NaN           NaN   
1                  NaN             NaN                 NaN           NaN   
2                  NaN             NaN                 NaN           NaN   
3                  NaN             NaN                 NaN           NaN   
4                  NaN             NaN                 NaN           NaN   

   Unnamed: 20  
0          NaN  
1          NaN  
2          NaN  
3          NaN  
4          NaN  

[5 rows x 21 columns]

# Qualité des données

Objectifs :
* Nettoyer les données

**EDSDATA**

In [ ]:
EDSData.tail()

Country Name Country Code  \
886925     Zimbabwe          ZWE   
886926     Zimbabwe          ZWE   
886927     Zimbabwe          ZWE   
886928     Zimbabwe          ZWE   
886929     Zimbabwe          ZWE   

                                           Indicator Name  \
886925  Youth illiterate population, 15-24 years, male...   
886926  Youth literacy rate, population 15-24 years, b...   
886927  Youth literacy rate, population 15-24 years, f...   
886928  Youth literacy rate, population 15-24 years, g...   
886929  Youth literacy rate, population 15-24 years, m...   

              Indicator Code  1970  1971  1972  1973  1974  1975  ...  2060  \
886925      UIS.LP.AG15T24.M   NaN   NaN   NaN   NaN   NaN   NaN  ...   NaN   
886926     SE.ADT.1524.LT.ZS   NaN   NaN   NaN   NaN   NaN   NaN  ...   NaN   
886927  SE.ADT.1524.LT.FE.ZS   NaN   NaN   NaN   NaN   NaN   NaN  ...   NaN   
886928  SE.ADT.1524.LT.FM.ZS   NaN   NaN   NaN   NaN   NaN   NaN  ...   NaN   
886929  SE.ADT.1524.LT.MA.ZS   NaN   NaN   NaN   NaN   NaN   NaN  ...   NaN   

        2065  2070  2075  2080  2085  2090  2095  2100  Unnamed: 69  
886925   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN          NaN  
886926   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN          NaN  
886927   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN          NaN  
886928   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN          NaN  
886929   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN          NaN  

[5 rows x 70 columns]

==> Les 4 premieres colonne contiennent des informations qualitatif (Nom du pays, code du pays, nom de l'indicateur, code de l'indicateur). Puis les autres correpondent au valeur associé en fonction des années. La dernière est nommé Unnamed : 69 (sens inconnu). 886930 lignes pour 70 colonnes

1°/ Erreurs de type

In [ ]:
EDSData.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 886930 entries, 0 to 886929
Data columns (total 70 columns):
 #   Column          Non-Null Count   Dtype  
---  ------          --------------   -----  
 0   Country Name    886930 non-null  object 
 1   Country Code    886930 non-null  object 
 2   Indicator Name  886930 non-null  object 
 3   Indicator Code  886930 non-null  object 
 4   1970            72288 non-null   float64
 5   1971            35537 non-null   float64
 6   1972            35619 non-null   float64
 7   1973            35545 non-null   float64
 8   1974            35730 non-null   float64
 9   1975            87306 non-null   float64
 10  1976            37483 non-null   float64
 11  1977            37574 non-null   float64
 12  1978            37576 non-null   float64
 13  1979            36809 non-null   float64
 14  1980            89122 non-null   float64
 15  1981            38777 non-null   float64
 16  1982            37511 non-null   float64
 17  1983      

Rien ne semble est problèmatique avec les types, par contre la colonne Unnamed : 69 ne correspond à rien

In [9]:
EDSData_copy = EDSData.copy()

In [10]:
EDSData = EDSData.drop(columns=["Unnamed: 69"])
EDSData.head()

Country Name Country Code  \
0   Arab World          ARB   
1   Arab World          ARB   
2   Arab World          ARB   
3   Arab World          ARB   
4   Arab World          ARB   

                                      Indicator Name  Indicator Code  \
0  Adjusted net enrolment rate, lower secondary, ...      UIS.NERA.2   
1  Adjusted net enrolment rate, lower secondary, ...    UIS.NERA.2.F   
2  Adjusted net enrolment rate, lower secondary, ...  UIS.NERA.2.GPI   
3  Adjusted net enrolment rate, lower secondary, ...    UIS.NERA.2.M   
4  Adjusted net enrolment rate, primary, both sex...     SE.PRM.TENR   

        1970       1971       1972       1973       1974      1975  ...  2055  \
0        NaN        NaN        NaN        NaN        NaN       NaN  ...   NaN   
1        NaN        NaN        NaN        NaN        NaN       NaN  ...   NaN   
2        NaN        NaN        NaN        NaN        NaN       NaN  ...   NaN   
3        NaN        NaN        NaN        NaN        NaN       NaN  ...   NaN   
4  54.822121  54.894138  56.209438  57.267109  57.991138  59.36554  ...   NaN   

   2060  2065  2070  2075  2080  2085  2090  2095  2100  
0   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN  
1   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN  
2   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN  
3   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN  
4   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN  

[5 rows x 69 columns]

2°/ Valeurs manquantes

In [11]:
EDSData.isnull().sum()

Country Name           0
Country Code           0
Indicator Name         0
Indicator Code         0
1970              814642
                   ...  
2080              835494
2085              835494
2090              835494
2095              835494
2100              835494
Length: 69, dtype: int64

In [16]:
pd.set_option("display.max_columns", None)
missing_percent = EDSData.isnull().mean()*100

pd.set_option("display.max_rows", None)
missing_percent.sort_values(ascending=False)

2017              99.983877
2016              98.144160
1971              95.993258
1973              95.992356
1972              95.984012
1974              95.971497
1979              95.849842
1976              95.773849
1982              95.770692
1989              95.767422
1977              95.763589
1978              95.763364
1983              95.663694
1988              95.653321
1984              95.647233
1987              95.643286
1981              95.627953
1986              95.560867
2045              94.200670
2100              94.200670
2030              94.200670
2095              94.200670
2090              94.200670
2025              94.200670
2080              94.200670
2085              94.200670
2050              94.200670
2035              94.200670
2040              94.200670
2065              94.200670
2070              94.200670
2075              94.200670
2060              94.200670
2055              94.200670
2020              94.200670
1970              91.849639
1997              91.718287
1991              91.607342
1992              91.482642
1993              91.454455
1996              91.340128
1994              91.266278
1998              90.426076
1975              90.156382
1980              89.951631
1985              89.819264
2014              87.170464
1999              86.601085
2001              86.074549
2002              85.996076
1990              85.973527
2004              85.476419
2003              85.301771
2015              85.223411
1995              85.189248
2008              84.848071
2007              84.522792
2013              84.496071
2006              84.180037
2009              83.977541
2011              83.537370
2012              83.396209
2000              80.080051
2005              79.242105
2010              72.665036
Indicator Code     0.000000
Indicator Name     0.000000
Country Code       0.000000
Country Name       0.000000
dtype: float64

Objectif : supprimer les lignes qui sont toutes en NaN

In [33]:
# Sélection des colonnes 4 à 67
cols = EDSData.columns[4:68]

# Supprimer lignes où toutes ces colonnes sont NaN
EDSData_clean = EDSData.dropna(subset=cols, how="all")
EDSData_clean.info()

<class 'pandas.core.frame.DataFrame'>
Index: 357405 entries, 4 to 886929
Data columns (total 69 columns):
 #   Column          Non-Null Count   Dtype  
---  ------          --------------   -----  
 0   Country Name    357405 non-null  object 
 1   Country Code    357405 non-null  object 
 2   Indicator Name  357405 non-null  object 
 3   Indicator Code  357405 non-null  object 
 4   1970            72288 non-null   float64
 5   1971            35537 non-null   float64
 6   1972            35619 non-null   float64
 7   1973            35545 non-null   float64
 8   1974            35730 non-null   float64
 9   1975            87306 non-null   float64
 10  1976            37483 non-null   float64
 11  1977            37574 non-null   float64
 12  1978            37576 non-null   float64
 13  1979            36809 non-null   float64
 14  1980            89122 non-null   float64
 15  1981            38777 non-null   float64
 16  1982            37511 non-null   float64
 17  1983           

In [34]:
missing_percent_clean = EDSData_clean.isnull().mean()*100

pd.set_option("display.max_rows", None)
missing_percent_clean.sort_values(ascending=False)

2017              99.959989
2016              95.394580
1971              90.056938
1973              90.054700
1972              90.033995
1974              90.002938
1979              89.701039
1976              89.512458
1982              89.504624
1989              89.496510
1977              89.486997
1978              89.486437
1983              89.239099
1988              89.213357
1984              89.198248
1987              89.188456
1981              89.150404
1986              88.983926
2045              85.608483
2100              85.608483
2030              85.608483
2095              85.608483
2090              85.608483
2025              85.608483
2080              85.608483
2085              85.608483
2050              85.608483
2035              85.608483
2040              85.608483
2065              85.608483
2070              85.608483
2075              85.608483
2060              85.608483
2055              85.608483
2020              85.608483
1970              79.774206
1997              79.448245
1991              79.172927
1992              78.863474
1993              78.793526
1996              78.509814
1994              78.326548
1998              76.241519
1975              75.572250
1980              75.064143
1985              74.735664
2014              68.162449
1999              66.749486
2001              65.442845
2002              65.248108
1990              65.192149
2004              63.958534
2003              63.525133
2015              63.330675
1995              63.245898
2008              62.399239
2007              61.592031
2013              61.525720
2006              60.741456
2009              60.238945
2011              59.146626
2012              58.796323
2000              50.567004
2005              48.487570
2010              32.166030
Indicator Code     0.000000
Indicator Name     0.000000
Country Code       0.000000
Country Name       0.000000
dtype: float64

Maintenant que les lignes full NaN sont supprimées j'observe que les annèes cette répartition :
* colonne quasiment vides (>85%) : 2017, 2016, 1971 à 1986
* Colonnes avec beaucoup de NaN mais encore acceptable (50-85%) : 2045 à 2085, 1970 à 1998, 2000 à 2014
* Colonne plutot complètes (<50%) : 2025, 2010

